File: +ndi/+daq/+reader/+mfdaq/blackrock.m
% NDI_DAQREADER_MFDAQ_BLACKROCK - Device driver for Blackrock Microsystems NSx/NEV file format
%
% This class reads data from Blackrock Microsystems NSx/NEV file format.
%
% Blackrock Microsystems: https://www.blackrockmicro.com/
%
%

classdef blackrock < ndi.daq.reader.mfdaq
    properties

    end % properties

    methods
        function obj = blackrock(varargin)
            % ndi.daq.reader.mfdaq.blackrock - Create a new NDI_DEVICE_MFDAQ_BLACKROCK object
            %
            %  D = ndi.daq.reader.mfdaq.blackrock()
            %
            %  Creates a new ndi.daq.reader.mfdaq.blackrock object
            %
            obj = obj@ndi.daq.reader.mfdaq(varargin{:})
        end

        function channels = getchannelsepoch(ndi_daqreader_mfdaq_blackrock_obj, epochfiles)
            % GETCHANNELSEPOCH - List the channels that are available on this Blackrock device for a given set of files
            %
            %  CHANNELS = GETCHANNELSEPOCH(NDI_DAQREADER_MFDAQ_BLACKROCK_OBJ, EPOCHFILES)
            %
            %  Returns the channel list of acquired channels in this session
            %
            % CHANNELS is a structure list of all channels with fields:
            % -------------------------------------------------------
            % 'name'             | The name of the channel (e.g., 'ai1')
            % 'type'             | The type of data stored in the channel
            %                    |    (e.g., 'analogin', 'digitalin', 'image', 'timestamp')
            %

            [ns_h,nev_h,headers] = read_blackrock_headers(ndi_daqreader_mfdaq_blackrock_obj, epochfiles);
            % to do: need to search nev file
            channels = vlt.data.emptystruct('name','type');
            for i=1:numel(ns_h.MetaTags.ChannelID)
                newchannel.type = 'analog_in';
                newchannel.name = ['ai' int2str(ns_h.MetaTags.ChannelID(i))];
                channels(end+1) = newchannel;
            end;

        end % getchannels()

        function [b,msg] = verifyepochprobemap(ndi_daqreader_mfdaq_blackrock_obj, epochprobemap, epochfiles)
            % VERIFYEPOCHPROBEMAP - Verifies that an EPOCHPROBEMAP is compatible with a given device and the data on disk
            %
            %   B = VERIFYEPOCHPROBEMAP(NDI_DAQREADER_MFDAQ_BLACKROCK_OBJ, EPOCHPROBEMAP, EPOCHFILES)
            %
            % Examines the NDI_EPOCHPROBEMAP_DAQREADER EPOCHPROBEMAP and determines if it is valid for the given device
            % with epoch files EPOCHFILES.
            %
            % See also: ndi.daq.reader, NDI_EPOCHPROBEMAP_DAQREADER
            b = 1;
            msg = '';
            % UPDATE NEEDED
        end

        function data = readchannels_epochsamples(ndi_daqreader_mfdaq_blackrock_obj, channeltype, channel, epochfiles, s0, s1)
            %  READCHANNELS_EPOCHSAMPLES - read the data based on specified channels
            %
            %  DATA = READCHANNELS_EPOCHSAMPLES(MYDEV, CHANNELTYPE, CHANNEL, EPOCHFILES ,S0, S1)
            %
            %  CHANNELTYPE is the type of channel to read (cell array of strings, one per channel)
            %
            %  CHANNEL is a vector of the channel numbers to read, beginning from 1
            %
            %  EPOCH is set of epoch files
            %
            %  DATA is the channel data (each column contains data from an individual channel)
            %
            [nev_files, nsv_files] = ndi.daq.reader.mfdaq.blackrock.filenamefromepochfiles(epochfiles);

            [ns_h,nev_h,headers] = ndi_daqreader_mfdaq_blackrock_obj.read_blackrock_headers(epochfiles);

            uniquechannel = unique(channeltype);
            if numel(uniquechannel)~=1
                error(['Only one type of channel may be read per function call at present.']);
            end

            data = [];

            if s0 < 1
                s0 = 1; % in Blackrock, the first sample is always 1
            elseif s0 > ns_h.MetaTags.DataPoints
                s0 = ns_h.MetaTags.DataPoints;
            end;

            if s1 > ns_h.MetaTags.DataPoints
                s1 = ns_h.MetaTags.DataPoints;
            end;

            if s1 < s0
                data = zeros(0, numel(channel));
                return;
            end;

            if strcmp(vlt.data.celloritem(channeltype,1),'ai')
                for i=1:numel(channel)
                    ns_out = openNSx(nsv_files{1},'read','precision','double','uV','sample',...
                        ['t:' int2str(s0) ':' int2str(s1)], ['c:' int2str(channel(i))]);
                    if ~isstruct(ns_out)
                        error(['No data read from channel ' int2str(i) ' of blackrock record.']);
                    end;
                    data = cat(2,data,ns_out.Data');
                end;

            elseif strcmp(vlt.data.celloritem(channeltype,1),'time')
                data = cat(1,data,ns_h.MetaTags.Timestamp+((s0:s1)'-1)*1./ns_h.MetaTags.SamplingFreq);
            end;

        end % readchannels_epochsamples

        function sr = samplerate(ndi_daqreader_mfdaq_blackrock_obj, epochfiles, channeltype, channel)
            % SAMPLERATE - GET THE SAMPLE RATE FOR SPECIFIC EPOCH AND CHANNEL
            %
            % SR = SAMPLERATE(DEV, EPOCHFILES, CHANNELTYPE, CHANNEL)
            % CHANNELTYPE can be either a string or a cell array of
            % strings the same length as the vector CHANNEL.
            % If CHANNELTYPE is a single string, then it is assumed that
            % that CHANNELTYPE applies to every entry of CHANNEL.
            %
            % SR is the list of sample rate from specified channels
            %
            sr = [];

            [ns_h,nev_h,headers] = read_blackrock_headers(ndi_daqreader_mfdaq_blackrock_obj, epochfiles, channeltype, channel);

            for i=1:numel(channel)
                ct = vlt.data.celloritem(channeltype,i);
                if strcmpi(ct,'ai') | strcmpi(ct,'time')
                    sr(i) = ns_h.MetaTags.SamplingFreq;
                else
                    error(['At present, do not know how to handle Blackrock Micro channels of type ' ct '.']);
                end;
            end;
        end % samplerate()

        function [ns_h,nev_h,headers] = read_blackrock_headers(ndi_daqreader_mfdaq_blackrock_obj, epochfiles, channeltype, channels)
            % READ_BLACKROCK_HEADERS - read information from Blackrock Micro header files
            %
            % [NS_H, NEV_H, HEADERS] = READ_BLACKROCK_HEADERS(NDI_DAQREADER_MFDAQ_BLACKROCK_OBJ, EPOCHFILES, [CHANNELTYPE, CHANNELS])
            %
            [nev_files, nsv_files] = ndi.daq.reader.mfdaq.blackrock.filenamefromepochfiles(epochfiles);
            if ~isempty(nsv_files{1})
                ns_h = openNSx(nsv_files{1},'noread');
            else
                ns_h = [];
            end;
            if ~isempty(nev_files{1})
                nev_h = openNEV(nev_files{1},'noread');
                nev_h = [];
            end;

            headers.ns_rate = [];
            if ~isempty(ns_h)
                headers.ns_rate = ns_h.MetaTags.SamplingFreq;
            end;
            headers.requestedchanneltype = [];
            headers.requestedchannelindexes= [];

            if nargin>=3
                for i=1:numel(channels)
                    ct = vlt.data.celloritem(channeltype,i);
                    if strcmpi(ct,'ai')
                        if isempty(ns_h)
                            error(['ai channels in Blackrock must be stored in .ns# files, but there is none.']);
                        end;
                        index = find(ns_h.MetaTags.ChannelID==channels(i));
                        if isempty(index)
                            error(['Channel ' int2str(channels(i)) ' not recorded.']);
                        else
                            headers.requestedchannelindexes(i) = index;
                            headers.requestedchanneltype(i) = 1; % ns==1, nev==2
                        end;
                    else
                        error(['At present, do not know how to handle Blackrock Micro channels of type ' ct '.']);
                    end;
                end;
            end;
        end; % read_blackrock_headers()

        function t0t1 = t0_t1(ndi_daqreader_mfdaq_blackrock_obj, epochfiles)
            % T0_T1 - return the t0_t1 (beginning and end) epoch times for an epoch
            %
            % T0T1 = T0_T1(NDI_EPOCHSET_OBJ, EPOCHFILES)
            %
            % Return the beginning (t0) and end (t1) times of the epoch EPOCH_NUMBER
            % in the same units as the ndi.time.clocktype objects returned by EPOCHCLOCK.
            %
            % The abstract class always returns {[NaN NaN]}.
            %
            % See also: ndi.time.clocktype, EPOCHCLOCK
            %
            [ns_h,nev_h,headers] = read_blackrock_headers(ndi_daqreader_mfdaq_blackrock_obj, epochfiles);
            % need to convert from duration of whole recording to time labels of the first and last sample
            % time of last sample = duration - 1/samplingfreq
            t0t1 = {[ns_h.MetaTags.Timestamp + [0 ns_h.MetaTags.DataDurationSec-1/ns_h.MetaTags.SamplingFreq]]};
            % developer note: in the Blackrock acquisition software, one can define a time offset; right now we aren't considering that
        end % t0t1

    end % methods

    methods (Static)  % helper functions

        function [nevfiles, nsvfiles] = filenamefromepochfiles(filename_array)
            % FILENAMEFROMEPOCHFILES - return the file name that corresponds to the NEV/NSV files
            %
            % [NEVFILES, NSVFILES] = FILENAMEFROMEPOCHFILES(FILENAME_ARRAY)
            %
            % Examines the list of filenames in FILENAME_ARRAY (cell array of full path file strings) and determines which
            % ones have the extension '.nev' (neuro event file) and which have the extension '.ns#', where # is a number, or the source
            % data files.
            %
            sv = ['.*\.ns\d\>'];
            tf_sv = vlt.string.strcmp_substitution(sv,filename_array,'UseSubstituteString',0);
            nsvfiles = filename_array(find(tf_sv));

            ne_search = ['.*\.nev\>'];
            tf_ne = vlt.string.strcmp_substitution(ne_search,filename_array,'UseSubstituteString',0);
            nevfiles = filename_array(find(tf_ne));

            if numel(nsvfiles)+numel(nevfiles) == 0
                error(['No .ns# or .nev files found.']);
            end;

            if numel(nsvfiles)>1
                error(['More than 1 NS# file in this file list; do not know what to do.']);
            end;
        end % filenamefromepoch

    end % methods (Static)
end

File: ndi-python/ndi/daq/reader/mfdaq/blackrock.py
"""
NDI DAQ Reader MFDAQ Blackrock - Device driver for Blackrock Microsystems NSx/NEV file format.

This class reads data from Blackrock Microsystems NSx/NEV file format.

Blackrock Microsystems: https://www.blackrockmicro.com/
"""

from typing import List, Tuple, Dict, Any, Optional
import re
import os
import numpy as np
from ..mfdaq import MFDAQReader


class Blackrock(MFDAQReader):
    """
    Blackrock reader for Blackrock Microsystems NSx/NEV file format.

    This class reads data from Blackrock Microsystems NSx (neural signal)
    and NEV (neural event) file formats.

    Examples:
        >>> from ndi.daq.reader.mfdaq import Blackrock
        >>> reader = Blackrock()
        >>> channels = reader.getchannelsepoch(epochfiles)
    """

    def __init__(self, *args):
        """
        Create a new Blackrock reader.

        Args:
            *args: Arguments passed to MFDAQReader constructor
        """
        super().__init__(*args)

    def getchannelsepoch(self, epochfiles: List[str]) -> List[Dict[str, Any]]:
        """
        List the channels that are available on this Blackrock device for a given set of files.

        Args:
            epochfiles: List of file paths for this epoch

        Returns:
            List of channel dictionaries with fields:
            - 'name': The name of the channel (e.g., 'ai1')
            - 'type': The type of data stored in the channel (e.g., 'analog_in')
        """
        ns_h, nev_h, headers = self.read_blackrock_headers(epochfiles)

        channels = []

        # Add analog input channels from NS file
        if ns_h is not None and 'MetaTags' in ns_h:
            if 'ChannelID' in ns_h['MetaTags']:
                for channel_id in ns_h['MetaTags']['ChannelID']:
                    channels.append({
                        'name': f'ai{channel_id}',
                        'type': 'analog_in'
                    })

        # TODO: Add NEV channels when NEV support is implemented

        return channels

    def verifyepochprobemap(self, epochprobemap: Any, epochfiles: List[str]) -> Tuple[bool, str]:
        """
        Verifies that an EPOCHPROBEMAP is compatible with a given device and the data on disk.

        Args:
            epochprobemap: NDI EpochProbeMap object
            epochfiles: List of file paths for this epoch

        Returns:
            Tuple of (is_valid, error_message)
        """
        # TODO: Implement verification when EpochProbeMap is fully defined
        return True, ''

    def readchannels_epochsamples(self, channeltype: Any, channel: List[int],
                                  epochfiles: List[str], s0: int, s1: int) -> np.ndarray:
        """
        Read the data based on specified channels.

        Args:
            channeltype: Type of channel to read (string or list of strings)
            channel: List of channel numbers to read (1-indexed)
            epochfiles: List of file paths for this epoch
            s0: Start sample (1-indexed)
            s1: End sample (1-indexed)

        Returns:
            Data array where each column contains data from an individual channel
        """
        nev_files, nsv_files = self.filenamefromepochfiles(epochfiles)
        ns_h, nev_h, headers = self.read_blackrock_headers(epochfiles)

        # Ensure channeltype is a list
        if not isinstance(channeltype, list):
            channeltype = [channeltype] * len(channel)

        # Check that all channels are the same type
        unique_types = list(set(channeltype))
        if len(unique_types) != 1:
            raise ValueError('Only one type of channel may be read per function call at present.')

        data = []

        # Clamp sample bounds
        if s0 < 1:
            s0 = 1
        if ns_h and 'MetaTags' in ns_h:
            if s0 > ns_h['MetaTags']['DataPoints']:
                s0 = ns_h['MetaTags']['DataPoints']
            if s1 > ns_h['MetaTags']['DataPoints']:
                s1 = ns_h['MetaTags']['DataPoints']

        if s1 < s0:
            return np.zeros((0, len(channel)))

        # Read based on channel type
        ct = channeltype[0]
        if ct == 'ai' or ct == 'analog_in':
            # Read analog input channels
            for ch in channel:
                ch_data = self._read_blackrock_channel(nsv_files[0], ch, s0, s1)
                if data == []:
                    data = ch_data.reshape(-1, 1)
                else:
                    data = np.concatenate([data, ch_data.reshape(-1, 1)], axis=1)

        elif ct == 'time' or ct == 'timestamp':
            # Generate time array
            if ns_h and 'MetaTags' in ns_h:
                timestamp = ns_h['MetaTags'].get('Timestamp', 0)
                sampling_freq = ns_h['MetaTags'].get('SamplingFreq', 30000)

                # Time of each sample
                time_values = timestamp + ((np.arange(s0, s1 + 1) - 1) * (1.0 / sampling_freq))
                data = time_values.reshape(-1, 1)

        return data

    def samplerate(self, epochfiles: List[str], channeltype: Any,
                  channel: List[int]) -> List[float]:
        """
        Get the sample rate for specific epoch and channel.

        Args:
            epochfiles: List of file paths for this epoch
            channeltype: Type of channel (string or list of strings)
            channel: List of channel numbers

        Returns:
            List of sample rates for each channel
        """
        ns_h, nev_h, headers = self.read_blackrock_headers(epochfiles, channeltype, channel)

        sr = []
        for i, ch in enumerate(channel):
            ct = channeltype[i] if isinstance(channeltype, list) else channeltype

            if ct in ['ai', 'analog_in', 'time', 'timestamp']:
                if ns_h and 'MetaTags' in ns_h:
                    sr.append(ns_h['MetaTags'].get('SamplingFreq', 30000.0))
                else:
                    sr.append(30000.0)  # Default Blackrock sampling rate
            else:
                raise ValueError(f'At present, do not know how to handle Blackrock Micro channels of type {ct}.')

        return sr

    def read_blackrock_headers(self, epochfiles: List[str],
                               channeltype: Optional[Any] = None,
                               channels: Optional[List[int]] = None) -> Tuple[Optional[Dict], Optional[Dict], Dict]:
        """
        Read information from Blackrock Micro header files.

        Args:
            epochfiles: List of file paths for this epoch
            channeltype: Optional channel type (for validation)
            channels: Optional list of channel numbers (for validation)

        Returns:
            Tuple of (ns_h, nev_h, headers) where:
            - ns_h: NSx header dictionary
            - nev_h: NEV header dictionary (None for now)
            - headers: Additional header info
        """
        nev_files, nsv_files = self.filenamefromepochfiles(epochfiles)

        ns_h = None
        nev_h = None

        # Read NSx header
        if nsv_files and nsv_files[0]:
            ns_h = self._read_nsx_header(nsv_files[0])

        # Read NEV header
        if nev_files and nev_files[0]:
            # TODO: Implement NEV header reading
            nev_h = None

        # Build headers structure
        headers = {
            'ns_rate': None,
            'requestedchanneltype': [],
            'requestedchannelindexes': []
        }

        if ns_h and 'MetaTags' in ns_h:
            headers['ns_rate'] = ns_h['MetaTags'].get('SamplingFreq', None)

        # Validate requested channels if provided
        if channeltype is not None and channels is not None:
            for i, ch in enumerate(channels):
                ct = channeltype[i] if isinstance(channeltype, list) else channeltype

                if ct in ['ai', 'analog_in']:
                    if not ns_h:
                        raise ValueError('ai channels in Blackrock must be stored in .ns# files, but there is none.')

                    # Find channel in header
                    channel_ids = ns_h['MetaTags'].get('ChannelID', [])
                    try:
                        index = channel_ids.index(ch)
                        headers['requestedchannelindexes'].append(index)
                        headers['requestedchanneltype'].append(1)  # ns==1, nev==2
                    except ValueError:
                        raise ValueError(f'Channel {ch} not recorded.')
                else:
                    raise ValueError(f'At present, do not know how to handle Blackrock Micro channels of type {ct}.')

        return ns_h, nev_h, headers

    def t0_t1(self, epochfiles: List[str]) -> List[List[float]]:
        """
        Return the t0_t1 (beginning and end) epoch times for an epoch.

        Args:
            epochfiles: List of file paths for this epoch

        Returns:
            List containing [t0, t1] pair in seconds
        """
        ns_h, nev_h, headers = self.read_blackrock_headers(epochfiles)

        if not ns_h or 'MetaTags' not in ns_h:
            return [[0, 0]]

        # Get timestamp and duration
        timestamp = ns_h['MetaTags'].get('Timestamp', 0)
        duration_sec = ns_h['MetaTags'].get('DataDurationSec', 0)
        sampling_freq = ns_h['MetaTags'].get('SamplingFreq', 30000)

        # Time of first sample
        t0 = timestamp

        # Time of last sample = timestamp + duration - 1/samplingfreq
        t1 = timestamp + duration_sec - 1.0 / sampling_freq

        return [[t0, t1]]

    # Static helper methods

    @staticmethod
    def filenamefromepochfiles(filename_array: List[str]) -> Tuple[List[str], List[str]]:
        """
        Return the file names that correspond to the NEV/NSV files.

        Args:
            filename_array: List of full path file strings

        Returns:
            Tuple of (nevfiles, nsvfiles) where:
            - nevfiles: List of .nev files (neuro event files)
            - nsvfiles: List of .ns# files (neural signal files)

        Raises:
            ValueError: If no .ns# or .nev files found, or multiple .ns# files
        """
        # Look for .ns# files (ns1, ns2, ns3, ns4, ns5, ns6)
        nsv_pattern = r'.*\.ns\d$'
        nsvfiles = []
        for filepath in filename_array:
            if re.search(nsv_pattern, filepath, re.IGNORECASE):
                nsvfiles.append(filepath)

        # Look for .nev files
        nev_pattern = r'.*\.nev$'
        nevfiles = []
        for filepath in filename_array:
            if re.search(nev_pattern, filepath, re.IGNORECASE):
                nevfiles.append(filepath)

        if len(nsvfiles) + len(nevfiles) == 0:
            raise ValueError('No .ns# or .nev files found.')

        if len(nsvfiles) > 1:
            raise ValueError('More than 1 NS# file in this file list; do not know what to do.')

        return nevfiles, nsvfiles

    # Private helper methods for file I/O (placeholders for external dependencies)

    def _read_nsx_header(self, filename: str) -> Dict[str, Any]:
        """
        Read Blackrock NSx header.

        Args:
            filename: Path to NSx file

        Returns:
            Header dictionary

        Note:
            This is a placeholder for the external openNSx function.
            TODO: Implement or import from Blackrock SDK when available.
        """
        # TODO: Implement actual NSx header reading
        # For now, return minimal structure
        return {
            'MetaTags': {
                'ChannelID': [],
                'DataPoints': 0,
                'DataDurationSec': 0,
                'SamplingFreq': 30000,
                'Timestamp': 0
            }
        }

    def _read_blackrock_channel(self, filename: str, channel: int,
                                s0: int, s1: int) -> np.ndarray:
        """
        Read Blackrock channel data.

        Args:
            filename: Path to NSx file
            channel: Channel number
            s0: Start sample (1-indexed)
            s1: End sample (1-indexed)

        Returns:
            Data array

        Note:
            This is a placeholder for the external openNSx function.
            TODO: Implement or import from Blackrock SDK when available.
        """
        # TODO: Implement actual NSx data reading
        # The MATLAB version calls:
        # ns_out = openNSx(filename,'read','precision','double','uV','sample',
        #                  ['t:' int2str(s0) ':' int2str(s1)], ['c:' int2str(channel)])
        return np.array([])

    def __repr__(self) -> str:
        """String representation."""
        return f"Blackrock(name='{self.name if hasattr(self, 'name') else ''}')"

    def __str__(self) -> str:
        """String representation."""
        return self.__repr__()
